### PR TITLE
Provenance data to project variables

### PR DESCRIPTION
**Pull request by ismailsunni**
### What does it fix?
* Ticket: #4303
* Funded by: DFAT
* Description: 
  - When user click output layer from analysis that has provenance, InaSAFE will set the value of the provenance in the project variables

screen shot 2017-07-25 at 10 31 28

cc @timlinux

### CODE REVIEW

File: safe/gui/widgets/dock.py [hunk 1]
```diff
@@ -1396,3 +1406,29 @@ def _validate_question_area(self):
...
+
+    def set_provenance_to_project_variables(self, provenances):
+        """Helper method to update / create provenance in project variables.
+
+        :param provenances: Keys and values from provenances.
+        :type provenances: dict
+        """
+        # QgsExpressionContextUtils.setProjectVariable(provenances)
+        for key, value in provenances.items():
+            if isinstance(value, (str, unicode, int, float)):
```
Comment (ismailsunni) [inline safe/gui/widgets/dock.py]: __[flake8]__
* [F821] undefined name 'unicode'

File: safe/gui/widgets/dock.py [hunk 2]
```diff
@@ -1396,3 +1405,44 @@ def _validate_question_area(self):
...
+    def set_provenance_to_project_variables(self, provenances):
+        """Helper method to update / create provenance in project variables.
+
+        :param provenances: Keys and values from provenances.
+        :type provenances: dict
+        """
+        # Remove old provenance data first
+        self.remove_provenance_project_variables()
+        for key, value in provenances.items():
+            if isinstance(value, (str, unicode, Number)):
```
Comment (ismailsunni) [inline safe/gui/widgets/dock.py]: __[flake8]__
* [F821] undefined name 'unicode'

File: safe/definitions/__init__.py [hunk 3]
```diff
@@ -26,6 +26,7 @@
 from safe.definitions.messages import *
 from safe.definitions.minimum_needs import *
 from safe.definitions.post_processors import *
+from safe.definitions.provenances import *
```
Comment (ismailsunni) [inline safe/definitions/__init__.py]: __[flake8]__
* [F403] 'from safe.definitions.provenances import *' used; unable to detect undefined names
* [F401] 'safe.definitions.provenances.*' imported but unused

File: safe/gui/widgets/dock.py [hunk 4]
```diff
@@ -57,6 +60,7 @@
     add_ordered_combo_item,
     is_keyword_version_supported,
 )
+from safe.metadata.utilities import serialize_dictionary
```
Comment (ismailsunni) [inline safe/gui/widgets/dock.py]: __[flake8]__
* [F401] 'safe.metadata.utilities.serialize_dictionary' imported but unused

File: safe/gui/widgets/dock.py [hunk 5]
```diff
@@ -1396,3 +1404,65 @@ def _validate_question_area(self):
...
+            """
+            if isinstance(value, (list, tuple, set)):
+                for element_index, element_value in enumerate(value):
+                    write_project_variable(
+                        '%s__%s' % (key, element_index), element_value)
+            elif isinstance(value, dict):
+                for dict_key, dict_value in value.items():
+                    write_project_variable(
+                        '%s__%s' % (key, dict_key), dict_value)
+            elif isinstance(value, (bool, str, unicode, Number)):
```
Comment (ismailsunni) [inline safe/gui/widgets/dock.py]: __[flake8]__
* [F821] undefined name 'unicode'

File: safe/gui/widgets/dock.py [hunk 6]
```diff
@@ -1396,3 +1403,65 @@ def _validate_question_area(self):
...
```
Comment (ismailsunni) [inline safe/gui/widgets/dock.py]: __[flake8]__
* [F821] undefined name 'unicode'

File: safe/gui/widgets/dock.py (hunk 5 above)
Comment (Gustry) [inline safe/gui/widgets/dock.py]: Can't you set Python 2 in this checker? Maybe it's using python3 to check this.

File: safe/metadata/constants.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
+# coding=utf-8
+
+"""Metadata Constants"""
```
Comment (Gustry) [inline safe/metadata/constants.py]: Not in safe/definitions?

**Comment by Gustry** [comment]
When you click outside of the group in the legend, it removes theses variables?
Why exposure_keywords and hazard_keywords are empty in your screenshot?

File: safe/definitions/__init__.py (hunk 3 above)
Comment (timlinux) [inline safe/definitions/__init__.py]: > from safe.definitions.provenances import *

s/provenances/provenance

**Comment by timlinux** [comment]
Some comments from me:

Entries like 'action_checklist' should be presented as plain text string or html markup, not JSON.

Impact Function Name: Should strip out the underscores

Some things like host name, python version, pyqt, qt, qgis may already be provided in the QGIS global variables - we should avoid duplicating them in project variables when already provided.

The main things I am missing here are references to the sub products:

* analysis_layer
* aggregation_impacted layer
* etc.

Having their file system paths / URI's will enable smart users to load that data dynamically and grab entries from it. I think we might provide some custom python QgsExpressions which make convenience helpers for getting content from these layers as @Gustry and I discussed in JK on our last site visit.

Basically everything else we provide here should be nicely formatted and ready to use to place into a layout (so use nice capitalization, remove 'foreign' looking stuff like underscores, use clean, logical naming etc. 

It would be also good to see if we can upstream a patch to make it easier to read and write project variables via the QGIS python API.

Lastly it will be great if you can confirm the behaviour of what happens when you click off a layer group and when you swap between analysis groups? 

Thanks for this, great work!

File: safe/definitions/__init__.py [hunk 8]
```diff
@@ -26,6 +26,7 @@
 from safe.definitions.messages import *
 from safe.definitions.minimum_needs import *
 from safe.definitions.post_processors import *
+from safe.definitions.provenance import *
```
Comment (ismailsunni) [inline safe/definitions/__init__.py]: __[flake8]__
* [F403] 'from safe.definitions.provenance import *' used; unable to detect undefined names
* [F401] 'safe.definitions.provenance.*' imported but unused

File: safe/gui/widgets/dock.py (hunk 5 above)
Comment (ismailsunni) [inline safe/gui/widgets/dock.py]: Hmm, I have checked in the SideCI. It seems we need to set the flake8 preference.
> You can customize the behaviour and some other options by adding a `tox.ini` or `setup.cfg` file to your repository.

I will do it later (not in this PR).

**Comment by ismailsunni** [comment]
Hi @Gustry 

> When you click outside of the group in the legend, it removes theses variables?

Yes, it does.

> Why exposure_keywords and hazard_keywords are empty in your screenshot?

Wrong screenshot :D, I have updated it. No longer empty.

**Comment by ismailsunni** [comment]
> Entries like 'action_checklist' should be presented as plain text string or html markup, not JSON.

Wrong screenshot. It's already done in this PR. See my new screenshot above. :D

> Impact Function Name: Should strip out the underscores

Fixed.

> Some things like host name, python version, pyqt, qt, qgis may already be provided in the QGIS global variables - we should avoid duplicating them in project variables when already provided.

Ok. I will use the same key in the provenance. It's a good way to review all provenance that we have. I have fixed it also.

> The main things I am missing here are references to the sub products:
> 
> analysis_layer
> aggregation_impacted layer
> etc.
> Having their file system paths / URI's will enable smart users to load that data dynamically and grab entries from it. I think we might provide some custom python QgsExpressions which make convenience helpers for getting content from these layers as @Gustry and I discussed in JK on our last site visit.

Ok, I will need to add it to the provenance. Because previously we get the URI from resolving the path.

> Basically everything else we provide here should be nicely formatted and ready to use to place into a layout (so use nice capitalization, remove 'foreign' looking stuff like underscores, use clean, logical naming etc.

Ok. I found out we don't use the `name` from `definition` often. I will fix it also. But I will still store the computer-friendly string in the provenance (so that we can get the definition).

> It would be also good to see if we can upstream a patch to make it easier to read and write project variables via the QGIS python API.

Yeah, I am planning to do it.

> Lastly it will be great if you can confirm the behaviour of what happens when you click off a layer group and when you swap between analysis groups?

For layer group, I need to set the behaviour. Currently, it only works when we select output layer.

When we swap between analysis group, it will update the project variable directly.

> Thanks for this, great work!

Thanks

File: safe/gui/widgets/dock.py [hunk 9]
```diff
@@ -1396,3 +1403,73 @@ def _validate_question_area(self):
...
```
Comment (ismailsunni) [inline safe/gui/widgets/dock.py]: __[flake8]__
* [F821] undefined name 'unicode'

File: safe/gui/tools/wizard/test/test_wizard_util.py [hunk 10]
```diff
@@ -5,7 +5,7 @@
 from safe.test.utilities import get_qgis_app
 QGIS_APP, CANVAS, IFACE, PARENT = get_qgis_app()
 
-from safe.gui.tools.wizard.wizard_utils import get_question_text
+from safe.gui.tools.wizard.utilities import get_question_text
```
Comment (ismailsunni) [inline safe/gui/tools/wizard/test/test_wizard_util.py]: __[flake8]__
* [E402] module level import not at top of file

File: safe/impact_function/impact_function.py [hunk 11]
```diff
@@ -71,6 +76,38 @@
     layer_purpose_exposure_summary_table,
     layer_purpose_profiling,
 )
+from safe.definitions.provenance import (
```
Comment (ismailsunni) [inline safe/impact_function/impact_function.py]: __[flake8]__
* [F401] 'safe.definitions.provenance.provenance_pyqt_version' imported but unused

File: safe/definitions/provenance.py [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+provenance_list = [
+    provenance_action_checklist,
+    provenance_aggregation_keywords,
+    provenance_aggregation_layer,
+    provenance_aggregation_layer_id,
+    provenance_analysis_extent,
+    provenance_analysis_question,
+    provenance_data_store_uri,
+    provenance_duration,
+    provenance_end_datetime,
```
Comment (Gustry) [inline safe/definitions/provenance.py]: not possible to put `start_datetime` here in this list?

File: safe/metadata/property/dictionary_property.py [hunk 13]
```diff
@@ -1,12 +1,14 @@
-# -*- coding: utf-8 -*-
+# coding=utf-8
+
+"""Dictionary property"""
```
Comment (Gustry) [inline safe/metadata/property/dictionary_property.py]: Full stop in docstring

File: safe/definitions/provenance.py [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+    'provenance_key': 'impact_function_name'
+}
+provenance_impact_function_title = {
+    'key': 'provenance_impact_function_title',
+    'name': tr('Impact Function Title'),
+    'provenance_key': 'impact_function_title'
+}
+provenance_inasafe_version = {
+    'key': 'provenance_inasafe_version',
+    'name': tr('Inasafe Version'),
```
Comment (ismailsunni) [inline safe/definitions/provenance.py]: Oops, I am planning to fix the case in this file, since I was generating it using script. :D 

Thanks for the catch :D

File: safe/definitions/provenance.py (hunk 12 above)
Comment (ismailsunni) [inline safe/definitions/provenance.py]: I already put it, it's alphabetical, so it's not next to each other.

**Comment by ismailsunni** [comment]
> QGIS is already providing qgis_os_name, qgis_version for instance.

I will keep it in the metadata provenance. But, I won't save it in the project variables.

Actually, we have a case where the analysis is coming from another QGIS version, and is opened in the different QGIS version.

**Comment by ismailsunni** [comment]
Hi @timlinux 
I am merging this PR first since I already make a big changes, and I want to avoid conflict later. I will make another PR to address the rest tasks:
- Add behaviour when layer group is selected
- Add more content to provenance for impact layer
